lib/spack/spack/detection/test.py

```diff
@@ -0,0 +1,142 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+"""Create and run mock e2e tests for package detection."""
+import contextlib
+import pathlib
+import tempfile
+from typing import Generator, List, NamedTuple
+
+import jinja2
+
+from llnl.util import filesystem
+
+import spack.repo
+import spack.spec
+from spack.util import spack_yaml
+
+from .path import by_executable
+
+
+class MockExecutables(NamedTuple):
+    """Mock executables to be used in detection tests"""
+
+    #: Relative paths for mock executables to be created
+    executables: List[str]
+    #: Shell script for the mock executable
+    script: str
+
+
+class ExpectedTestResult(NamedTuple):
+    """Data structure to model assertions on detection tests"""
+
+    #: Spec to be detected
+    spec: str
+
+
+class DetectionTest(NamedTuple):
+    """Data structure to construct detection tests by PATH inspection.
+
+    Packages may have a YAML file containing the description of one or more detection tests
+    to be performed. Each test creates a few mock executable scripts in a temporary folder,
+    and checks that detection by PATH gives the expected results.
+    """
+
+    pkg_name: str
+    layout: List[MockExecutables]
+    results: List[ExpectedTestResult]
+
+
+class Runner:
+    """Runs an external detection test"""
+
+    def __init__(self, *, test: DetectionTest, repository: spack.repo.RepoPath):
+        self.test = test
+        self.repository = repository
+        self.tmpdir = pathlib.Path(tempfile.mkdtemp())
+
+    def execute(self) -> List[spack.spec.Spec]:
+        """Executes a test and returns the specs that have been detected.
+
+        This function sets-up a test in a temporary directory, according to the prescriptions
+        in the test layout, then performs a detection by executables and returns the specs that
+        have been detected.
+        """
+        with self._mock_layout() as path_hints:
+            entries = by_executable(
+                [self.repository.get_pkg_class(self.test.pkg_name)], path_hints=path_hints
+            )
+            specs = set(x.spec for x in entries[self.test.pkg_name])
+        return list(specs)
+
+    @contextlib.contextmanager
+    def _mock_layout(self) -> Generator[List[str], None, None]:
+        hints, to_be_removed = set(), []
+        try:
+            for entry in self.test.layout:
+                exes = self._create_executable_scripts(entry)
+                to_be_removed.extend(exes)
```

Comment:
I think this could leave files behind if `self._create_executable_scripts(entry)` raises an error. We should refactor this so the list of files to remove is calculated first, then the attempt to create them. It probably costs an extra iteration over the set of tests, but the list isn't long enough for that to matter compared to the I/O times.